feat(noise): add support to ChaChaPoly encryption

s1fr0 opened the pull request:
This PR adds support to the ChaChaPoly symmetric cipher for encrypting Waku payload message contents. 

Specifically, it adds:
- Encryption/Decryption primitives
- Some unit tests

It partially addresses https://github.com/status-im/nim-waku/issues/881.

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 23925658 | [#1](https://ci.status.im/job/nim-waku/job/prs/job/linux/job/PR-930/1/) | 2022-03-30 21:30:05 | ~12 min | `linux` | [:package:`bin`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/wakunode2-220330-211745-239256-linux.bin) |

oskarth commented on waku/v2/protocol/waku_noise/noise.nim:
```diff
@@ -0,0 +1,76 @@
+## Nim-LibP2P
```
Leftover copy paste? I don't think we have this part in other nim files

Maybe just mention it was taken/inspired by nim-libp2p with brief link?

oskarth replied:
I'd also add a link to RFC

s1fr0 replied:
Yes, my fault with copy paste from my bigger WIP source! No problem adding links in the source, although this part doesn't directly come from libp2p noise implementation.

oskarth commented on waku/v2/protocol/waku_noise/noise.nim:
```diff
@@ -0,0 +1,76 @@
+    state: ChaChaPolyCipherState,
+    plaintext: openArray[byte]): ChaChaPolyCiphertext
+    {.noinit, raises: [Defect].} =
+  #TODO: add padding
```
Can we raise a Github issue for this maybe?

oskarth replied:
OK not to address straight away, as long as we capture it somewhere as it may impact security

s1fr0 replied:
Sure, in fact padding is a requirement we have for encrypted after-handshake transport messages. Here we're still adding vanilla ChaChaPoly encryption, so padding is not necessarily a requirement, but it would be nice to have it as an option later on.

staheri14 commented on waku/v2/protocol/waku_noise/noise.nim:
```diff
@@ -0,0 +1,75 @@
+  topics = "nim-waku noise"
+
+const
+  # Empty is a special value which indicates k has not yet been initialized.
```
what is `k` in the comment?

staheri14 commented on tests/v2/test_waku_noise.nim:
```diff
@@ -0,0 +1,6 @@
+{.used.}
+
+import
+  testutils/unittests,
```
the test file looks empty, is it intentional?

s1fr0 replied:
Yes it was intentionally left blank since this PR comes from a process of cherry picking commits to split the big WIP implementation into smaller PRs. All changes to test_waku_noise.nim can in fact be discarded from this PR, but if we quickly merge this and all the other noise related PR then would not be necessary.

jm-clius replied:
Not important to address in this PR, but in future I think it's preferable having each incremental PR be reasonable as a stand-alone that can be reviewed in its entirety. It makes sense to have open TODOs, of course, but since empty modules can't really be reviewed we increase the risk of having dead code in future that we forget about.

jm-clius reviewed (changes requested):
Looks good. Some minor comments about Nim conventions.

s1fr0 commented:
@jm-clius addressed comments in https://github.com/status-im/nim-waku/commit/36b1104ed0a4056448f00bccd2c4b4ef182a5c34, https://github.com/status-im/nim-waku/commit/9d12464f1d0aaa17ff99ec02c324ce710702f725 and https://github.com/status-im/nim-waku/commit/852d8d91c897a4a28ecb92eb6db21b8a33414ef3

I added a new test to test encryption/decryption (it was missing before) and the file is no more empty. Removed debug tools import used in libp2p implementation. Restyled code according to suggested Nim style guide.

jm-clius reviewed (approved):
LGTM, thanks!
